### PR TITLE
[BE] feat: Implement user authentication and error handling

### PR DESCRIPTION
**Pull request by julianntorres**
- Introduced User structure and password comparison functions using bcrypt.
- Established UserRepository interface to fetch user information from the database.
- Created Authenticate use case to validate login information.
- Developed HandleLogin handler for managing HTTP login requests, deserializing payloads, and responding based on authentication outcome.
- Enhanced with specific validations for cases like user not found and invalid credentials.

The login module is designed to help users securely sign into the system, offering a clean and straightforward way for users to authenticate. Here's a deeper dive into how the login system operates:

**API Overview**

The Login service uses a RESTful API, relying on a POST operation to handle user authentication.

The requests and responses with this service are structured in JSON format, and they are handled synchronously to ensure timely feedback for the users.

**API Components and Operation**

**1. URL Schema:**

The endpoint to access the login service is structured as:
`https://{server}:{port}/login`

**2. Request Body:**

When sending a request to the login service, the following information is required:
<html>
<body>

Field | Description | Example
-- | -- | --
username | The user's chosen name or ID | lionel_cuccitini
password | The user's password | G0AT10

</body>
</html>

**3. Response Body:**
After sending the request, the server will respond with information about the authentication status. The response body contains:
<html>
<body>

Field | Description | Example
-- | -- | --
response | Detailed feedback on the outcome | Authentication successful, User not found, Invalid password, Authentication failed, Error during authentication
status | A short status code indicating the result | OK, NOTFOUND, NOPASSWORD, ERROR

</body>
</html>

Some sample responses from the server might be:

For a successful login:
```
{
 "response": "Authentication successful",
 "status": "OK"
}

```
If the password doesn't match:
```
{
 "response": "Invalid password",
 "status": "NOPASSWORD"
}


```

If the username isn't found in the system:
```
{
 "response": "User not found",
 "status": "NOTFOUND"
}


```

This structure ensures a smooth user experience, providing clear and concise feedback about the authentication process. Whether the user enters wrong credentials or encounters a system error, they'll receive an appropriate and understandable response.

### CODE REVIEW

File: backend/.air.toml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+root = "."
```
Comment (eds000n) [inline backend/.air.toml:1]: why do we need this file?

File: backend/internal/login/infrastructure/loginHandler.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(statusCode)
+	jsonResponse := map[string]string{
+		"status":   status,
+		"response": response,
+	}
+	json.NewEncoder(w).Encode(jsonResponse)
+}
+
+func (lh *LoginHandler) HandleLogin(w http.ResponseWriter, r *http.Request) {
```
Comment (eds000n) [inline backend/internal/login/infrastructure/loginHandler.go:28]: we need a unit test for this

File: backend/internal/login/infrastructure/database.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	config "github.com/unmsmfisi-socialapplication/social_app"
+	"github.com/unmsmfisi-socialapplication/social_app/internal/login/application"
+	"github.com/unmsmfisi-socialapplication/social_app/internal/login/domain"
+)
+
+var db *sql.DB
+
+func InitDatabase() error {
+	var err error
+	db, err = sql.Open("postgres", config.LoadConfig().DBConnectionString)
```
Comment (eds000n) [inline backend/internal/login/infrastructure/database.go:16]: why is this `database` file inside the `/internal/login/intraestructure`? don't we need db connection for other parts of the application as well?

File: backend/.air.toml (hunk 1 above)
Comment (julianntorres) [reply backend/.air.toml:1]: > why do we need this file?

This file it's not for a proper feature of the social network, it's being used for live reload the code so we can see changes on real time with no need of rebooting server manually, as a way to improve the development experience.

Comment (eds000n) [reply backend/.air.toml:1]: have you discussed the usage of this tool as a team?

File: backend/internal/login/infrastructure/loginHandler.go (hunk 2 above)
Comment (julianntorres) [reply backend/internal/login/infrastructure/loginHandler.go:28]: > we need a unit test for this

I'm currently working on it

File: backend/internal/login/infrastructure/database.go (hunk 3 above)
Comment (julianntorres) [reply backend/internal/login/infrastructure/database.go:16]: That's right, i am working on this correction so we can set a global file which all the applications components can import and use it to connect and use the database

Comment (julianntorres) [reply backend/internal/login/infrastructure/database.go:16]: > why is this `database` file inside the `/internal/login/intraestructure`? don't we need db connection for other parts of the application as well?

That's right, i am working on this correction so we can set a global file which all the applications components can import and use it to connect and use the database

File: backend/.air.toml (hunk 1 above)
Comment (julianntorres) [reply backend/.air.toml:1]: > have you discussed the usage of this tool as a team?

I've discussed the usage of this tool with some members (whom are using it at this moment) of our team, but not all of them yet

File: backend/internal/login/infrastructure/loginHandler.go (hunk 2 above)
Comment (julianntorres) [reply backend/internal/login/infrastructure/loginHandler.go:35]: Done

File: backend/.air.toml (hunk 1 above)
Comment (julianntorres) [reply backend/.air.toml]: Air has been removed

File: backend/internal/login/infrastructure/database.go (hunk 3 above)
Comment (julianntorres) [reply backend/internal/login/infrastructure/database.go]: Done

**Review by AnthonyRicse98 — APPROVED** [review]

**Review by RicseDev — APPROVED** [review]

**Review by RicseDev — APPROVED** [review]

**Review by AnthonyRicse98 — APPROVED** [review]

**Comment by julianntorres** [comment]
There's a few extra modifications to do on this component

**Pull request merged**